rust/src/mkcomposefs.rs

```diff
@@ -0,0 +1,114 @@
+//! # Creating composefs images
+//!
+//! This code wraps `mkcomposefs`, supporting synthesizing a composefs
+//! from dump file entries.
+
+use std::fs::File;
+use std::io::Write;
+use std::process::Command;
+use std::sync::mpsc;
+
+use anyhow::{Context, Result};
+
+/// Configuration for `mkcomposefs`
+#[derive(Debug, Default)]
+pub struct Config {
+    digest_store: Option<String>,
+    min_version: Option<u32>,
+    max_version: Option<u32>,
+}
+
+impl Config {
+    fn to_args(&self) -> impl Iterator<Item = String> {
+        self.digest_store
+            .as_deref()
+            .map(|v| format!("--digest-store={v}"))
+            .into_iter()
+            .chain(self.min_version.map(|v| format!("--min-version={v}")))
+            .chain(self.max_version.map(|v| format!("--max-version={v}")))
+    }
+}
+
+/// Prepare a child process invocation of `mkcomposefs`.  It will accept
+/// serialized dumpfile lines on stdin, and write output to stdout.
+fn new_mkcomposefs_command(config: Config, output: File) -> Result<Command> {
+    let mut proc = Command::new("mkcomposefs");
+    proc.args(config.to_args())
+        .args(["--from-file", "-", "-"])
+        .stdin(std::process::Stdio::piped())
+        .stderr(std::process::Stdio::piped())
+        .stdout(std::process::Stdio::from(output));
+    Ok(proc)
+}
+
+/// Given the provided configuration and dumpfile entries, write a composefs metadata file to `output`.
+pub fn mkcomposefs(
+    config: Config,
+    entries: mpsc::Receiver<super::dumpfile::Entry<'_>>,
+    output: File,
+) -> Result<()> {
+    let mut cmd = new_mkcomposefs_command(config, output)?;
+    let mut proc = cmd.spawn().context("Spawning mkcomposefs")?;
+    // SAFETY: we set up stdin
+    let mut child_stdin = std::io::BufWriter::new(proc.stdin.take().unwrap());
+    std::thread::scope(|s| {
+        // Spawn a helper thread which handles writing to the child stdin, while the main
+        // thread handles reading from stderr (if any) and otherwise just being blocked in wait().
+        // The composefs subprocess itself writes to the output file.
+        let writer = s.spawn(move || -> anyhow::Result<()> {
+            for entry in entries {
+                writeln!(child_stdin, "{entry}")?;
+            }
+            // Flush and close child's stdin
+            drop(child_stdin.into_inner()?);
+            Ok(())
+        });
+        let r = proc.wait_with_output()?;
+        if !r.status.success() {
+            let stderr = String::from_utf8_lossy(&r.stderr);
+            let stderr = stderr.trim();
+            anyhow::bail!("mkcomposefs failed: {}: {stderr}", r.status)
+        }
+        // SAFETY: We shouldn't fail to join the thread
+        writer.join().unwrap()?;
+        anyhow::Ok(())
+    })
+}
+
+#[test]
+fn test_mkcomposefs() -> Result<()> {
```

Comment:
I think this test is pretty neat: we're verifying we can round-trip everything through the Rust parser and serialization of the dump format, through `mkcomposefs` into an EROFS, and back out via `composefs-info dump`.